File: Python/intersection_of_two_arrays.py
# https://leetcode.com/problems/intersection-of-two-arrays/description/
from utils.examination import Examiner, Testcase


class Solution:
    def intersection(self, nums1: list[int], nums2: list[int]) -> list[int]:
        intersecteds: list[int] = []
        seen: dict[int, int] = {}
        for num in nums1:
            seen[num] = 0
        seen2: dict[int, int] = {}
        for num in nums2:
            if num in seen and not num in seen2:
                seen2[num] = 0
                intersecteds.append(num)
        return intersecteds

def examine()->None:
    Examiner(Solution(), 'intersection').run((
        Testcase(([], []), []),
        Testcase(([1,2,2,1], [2,2]), [2]),
        Testcase(([4,9,5], [9,4,9,8,4]), [9,4]),
    ))

File: TypeScript/solutions/intersection-of-two-arrays.ts
// https://leetcode.com/problems/intersection-of-two-arrays/description/
import { Examiner, Testcase } from '../utils';


function intersection(nums1: number[], nums2: number[]): number[] {
    const seen = new Set(nums1);
    const seen2 = new Set<number>();
    const result: number[] = [];
    for (const num of nums2) {
        if (seen.has(num) && !seen2.has(num)) { seen2.add(num); result.push(num); }
    }
    return result;
}

export function examine(): void {
    new Examiner(intersection).run([
        new Testcase([[], []], []),
        new Testcase([[1, 2, 2, 1], [2, 2]], [2]),
        new Testcase([[4, 9, 5], [9, 4, 9, 8, 4]], [9, 4]),
    ]);
}
